### ISSUE TITLE
想请教一下损失函数

### ISSUE THREAD

**Issue by XYS-999** [QinYang79/RDE#17]
作者您好，我想请教一下关于损失函数的问题。
1.我观察到以下代码，是否说明，论文只可以选择使用TAL，TRL，InfoNCE，SDM损失的其中一种，即使在option文件中指定了loss name为TAL+id+mlm，实际上也只使用了tal损失，相当于id和mlm是无效的。
因为我发现loss name为TAL+id+mlm或者TAL时，损失值是一样的。
`def compute_per_loss(image_features, text_features, pid, tau=0.02, margin=0.2, loss_type='TAL', logit_scale=50):
    
    # # normalized features
    image_norm = image_features / image_features.norm(dim=-1, keepdim=True)
    text_norm = text_features / text_features.norm(dim=-1, keepdim=True)
    scores = text_norm @ image_norm.t()

    if 'TAL' in loss_type:
        per_loss = compute_TAL_per(scores, pid, tau, margin=margin)
    elif 'TRL' in loss_type:
        per_loss = compute_TRL_per(scores, pid, tau=tau, margin=margin)
    elif 'InfoNCE' in loss_type:
        per_loss = compute_InfoNCE_per(scores, logit_scale)
    elif 'SDM' in loss_type:
        per_loss = compute_sdm_per(scores, pid, logit_scale)
    else:
        exit()

    return per_loss, scores.diag()
`
2.我使用损失为TAL+id+mlm，批量改为64，noisy_rate设为0.0，其余参数不变的情况，在CUHK数据集上，只跑到了71.93，与论文中的75.94相差很多，请问批量大小的影响很大吗？或者是其他什么参数没设置对？

**Comment by QinYang79** [comment]
> 作者您好，我想请教一下关于损失函数的问题。 1.我观察到以下代码，是否说明，论文只可以选择使用TAL，TRL，InfoNCE，SDM损失的其中一种，即使在option文件中指定了loss name为TAL+id+mlm，实际上也只使用了tal损失，相当于id和mlm是无效的。 因为我发现loss name为TAL+id+mlm或者TAL时，损失值是一样的。 `def compute_per_loss(image_features, text_features, pid, tau=0.02, margin=0.2, loss_type='TAL', logit_scale=50):
> 
> ```
> # # normalized features
> image_norm = image_features / image_features.norm(dim=-1, keepdim=True)
> text_norm = text_features / text_features.norm(dim=-1, keepdim=True)
> scores = text_norm @ image_norm.t()
> 
> if 'TAL' in loss_type:
>     per_loss = compute_TAL_per(scores, pid, tau, margin=margin)
> elif 'TRL' in loss_type:
>     per_loss = compute_TRL_per(scores, pid, tau=tau, margin=margin)
> elif 'InfoNCE' in loss_type:
>     per_loss = compute_InfoNCE_per(scores, logit_scale)
> elif 'SDM' in loss_type:
>     per_loss = compute_sdm_per(scores, pid, logit_scale)
> else:
>     exit()
> 
> return per_loss, scores.diag()
> ```
> 
> ` 2.我使用损失为TAL+id+mlm，批量改为64，noisy_rate设为0.0，其余参数不变的情况，在CUHK数据集上，只跑到了71.93，与论文中的75.94相差很多，请问批量大小的影响很大吗？或者是其他什么参数没设置对？

1. If you want to use id+mlm, you need to add the corresponding module yourself.
2. The weights and training logs have been released at [here](https://huggingface.co/Yangsss/RDE) and can be reproduced by running **run_rde.sh**.
3. You can check [here](https://github.com/QinYang79/RDE/issues/8) to see if it affects your training.
4. If you still cannot reproduce the results, it is recommended to attach your training log to help you troubleshoot the problem.

**Comment by XYS-999** [comment]
[train_log.txt](https://github.com/user-attachments/files/18823057/train_log.txt)
这是我运行的log，请问有什么问题呢

**Comment by QinYang79** [comment]
> [train_log.txt](https://github.com/user-attachments/files/18823057/train_log.txt) 这是我运行的log，请问有什么问题呢

 img_aug=False
 txt_aug=False

**Comment by XYS-999** [comment]
非常感谢，目前效果有所提升
